Add tutorial on how to build a reduced basis (#1013)
pymor/pymor: docs/source/tutorial-rb.rst

In pyMOR the set :math:`\mathcal{P}` is called the |ParameterSpace|.

How to read this formula? For each candidate reduced space :math:`V_N` we
look at all possible |parameter values| `\mu` and compute the best-approximation
error in :math:`V_N` (the second infimum). The supremum over the infimum
is thus the worst-case best-approximation error over all |parameter values| of
interest. Now we take the infimum of the worst-case best-approximation errors
over all possible reduced spaces of dimension at most :math:`N`, and this is
:math:`d_N`.

So whatever reduced space of dimension :math:`N` we pick, we
will always find a :math:`\mu` for which the best-approximation error in our
space is at least :math:`d_N`. Reduced basis methods aim at constructing
spaces :math:`V_N` for which the worst-case best-approximation error is as
close to :math:`d_N` as possible.

However, we will only find a good :math:`V_N` of small dimension
:math:`N` if the values :math:`d_N` decrease quickly for growing
:math:`N`. It can be shown that this is the case as soon as :math:`u(\mu)`
analytically depends on :math:`\mu`, which is the true for many problems

> of interest. More precisely, it can be shown that there are constants

Reviewer: What about giving a reference here?

Author: Done.

with pyMOR and study their error decay.

Model setup
-----------

First we need to define a |Model| and a |ParameterSpace| for which we want
to build a reduced basis. We choose here the standard
:meth:`thermal block <pymor.analyticalproblems.thermalblock.thermal_block_problem>` benchmark
problem shipped with pyMOR (see :doc:`getting_started`). However, any pyMOR
|Model| can be used, except for Section :ref:`weakgreedy` where
some more assumptions have to be made on the |Model|.

First we import everything we need:

.. jupyter-execute::

    import numpy as np
    from pymor.basic import *

Then we build a 3-by-3 thermalblock problem that we discretize using pyMOR's
:mod:`builtin discretizers <pymor.discretizers.builtin>` (see
:doc:`tutorial01` for an introduction to pyMOR's discretization toolkit).

.. jupyter-execute::

    problem = thermal_block_problem((3,3))
    fom, _ = discretize_stationary_cg(problem, diameter=1/100)

Next, we need to define a |ParameterSpace| of |parameter values| for which
the solutions of `fom` should be approximated by the reduced basis. We do
this by calling the :meth:`~pymor.parameters.base.Parameters.space` method

> of the |parameters| of `m`:

Reviewer: ```suggestion
of the |parameters| of `fom`:
```

.. jupyter-execute::

    parameter_space = fom.parameters.space(0.0001, 1.)

Here, `0.0001` and `1.` are the common lower and upper bounds for the
individual components of all |parameters| of `fom`. In our case `fom`
expects a single parameter `'diffusion'` of 9 values:

.. jupyter-execute::

   fom.parameters

If `fom` were to depend on multiple |parameters|, we could also call the
:meth:`~pymor.parameters.base.Parameters.space` method with a dictionary
of lower and upper bounds per parameter.

The main use of |ParameterSpaces| in pyMOR is that they allow to easily sample
|parameter values| from their domain using the methods
:meth:`~pymor.parameters.base.ParameterSpace.sample_uniformly` and
:meth:`~pymor.parameters.base.ParameterSpace.sample_randomly`.

Computing the snapshot data
---------------------------

Reduced basis methods are snapshot-based, which means that they build

> the reduced space as a linear subspace of the linear span of solutions

Reviewer: I think the point @sdrave wants to make here is that a bunch of solutions might be computed beforehand, but only some of them are actually used to create the reduced basis/space. Of course, you are right and it is not absolutely necessary to write it that way, but I think it's a matter of how to read this part.

> the reduced space as a linear subspace of the linear span of solutions

Author: In particular, for POD it is generally not true, that the reduced space is the span of snapshots.

> the reduced space as a linear subspace of the linear span of solutions

Reviewer 2: Aha, ok, I get the meaning now.

> of the `fom` for certain |parameter values|. The easiest

Reviewer: ```suggestion
of the full-order model for certain |parameter values|. The easiest
```
Sounds better for me.

corresponding |VectorSpace|. Empty arrays are created using the
:meth:`~pymor.vectorarrays.interface.VectorSpace.empty` method. But what
is the right |VectorSpace|? All :meth:`solutions <pymor.models.interface.Model.solve>`
of a |Model| belong to its :attr:`~pymor.models.interface.Model.solution_space`,
so we write:

.. jupyter-execute::

    U = fom.solution_space.empty()
    for mu in training_set:
        U.append(fom.solve(mu))

Note that `fom.solve` returns a |VectorArray| containing a single vector.
This entire array (of one vector) is then appended to the `U` array.
pyMOR has no notion of single vectors, we only speak of |VectorArrays|.

What exactly is a |VectorSpace|? A |VectorSpace| in pyMOR holds all
information necessary to build |VectorArrays| containing vector objects
of a certain type. In our case we have

.. jupyter-execute::

   fom.solution_space

which means that the created |VectorArrays| will internally hold
|NumPy arrays| for data storage. The number is the dimension of the
vector. We have here a |NumpyVectorSpace| because pyMOR's builtin
discretizations are built around the NumPy/SciPy stack. If `fom` would

> represent a |Model| living in an external PDE solver, we would have

Reviewer: Maybe reference to the tutorial for external PDE solvers?

Author: Can't do that yet ..

Reviewer: You're right, maybe we could keep that in mind and add a reference later on.

:math:`u(\mu)` and its orthogonal projection will be the best-approximation
error.

So let :math:`v_{proj}` be the orthogonal projection of :math:`v` onto the
linear space spanned by the basis vectors :math:`u_i,\ i=1, \ldots, N`.
By definition this means that :math:`v - v_{proj}` is orthogonal to
all :math:`u_i`:

.. math:: (v - v_{proj}, u_i) = 0, \qquad i = 1, \ldots, N.

Let :math:`\lambda_j`, :math:`j = 1, \ldots, N` be the coefficients of
:math:`v_{proj}` with respect to  this basis, i.e.
:math:`\sum_{j=1}^N \lambda_j u_j = v_{proj}`. Then:

.. math::  \sum_{j=1}^N \lambda_j (u_j, u_i) = (v, u_i), \qquad i = 1, \ldots, N.

With :math:`G_{i,j} := (u_i, u_j)`,
:math:`R := [(v, u_1), \ldots, (v, u_N)]^T` and
:math:`\Lambda := [\lambda_1, \ldots, \lambda_N]^T`, we obtain the
linear equation system

.. math::  G \cdot \Lambda = R,

which determines :math:`\lambda_i` and, thus, :math:`v_{proj}`.

Let's assemble and solve this equation system using pyMOR to determine
the best-approximation error in `trivial_basis` for some test vector
`V` which we take as another random solution of our |Model|:

.. jupyter-execute::

    V = fom.solve(parameter_space.sample_randomly(1)[0])

> The matrix G of all inner products between vectors in `trivial_basis`

Reviewer: ```suggestion
The matrix :math:`G` of all inner products between vectors in `trivial_basis`
```

is a so called `Gramian matrix <https://en.wikipedia.org/wiki/Gramian_matrix>`_.

> Luckily, every ``VectorArray`` has a ``gramian`` method, which computes precisely

Reviewer: ```suggestion
Luckily, every |VectorArray| has a :meth:`~pymor.vectorarrays.interface.VectorArray.gramian` method, which computes precisely
```

right-hand side :math:`R`, we need to compute all (Euclidean) inner
products between the vectors in `trivial_basis` and (the single vector in)
`V`. For that, we can use the :meth:`~pymor.vectorarrays.interface.VectorArray.inner`
method:

.. jupyter-execute::

    R = trivial_basis.inner(V)

which will give us a :math:`25\times 1` |NumPy array| of all inner products.

.. jupyter-execute::
    :hide-code:
    :hide-output:

    assert R.shape == (25,1)

Now, we can use |NumPy| to solve the linear equation system:

.. jupyter-execute::

    lambdas = np.linalg.solve(G, R)

Finally, we need to form the linear combination

.. math:: \sum_{j=1}^N \lambda_j u_j = v_{proj}

using the :meth:`~pymor.vectorarrays.interface.VectorArray.lincomb` method
of `trivial_basis`. It expects row vectors of linear coefficients, but
`solve` returns column vectors, so we need to take the transpose:

.. jupyter-execute::

    V_proj = trivial_basis.lincomb(lambdas.T)

Let's look at `V`, `V_proj` and the difference of both. |VectorArrays| of
the same length can simply be subtracted, yielding a new array of the
differences:

.. jupyter-execute::

    fom.visualize((V, V_proj, V - V_proj),

>                   legend=('V', 'V_proj', 'best-approximation errr'),

Reviewer: ```suggestion
                  legend=('V', 'V_proj', 'best-approximation err'),
```

                  separate_colorbars=True)

As you can see, we already have a quite good approximation of `V` with
only 25 basis vectors.

Now, the Euclidean norm will just work fine in many cases.
However, when `fom` comes from a PDE, it will be usually not the norm

> we are interested in, and you may get poor results for problem with

Reviewer: ```suggestion
we are interested in, and you may get poor results for problems with
```

    `fom.h1_semi_product`.

To use `fom.h1_0_semi_product` as an inner product |Operator| for computing the
projection error, we can simply pass it as the optional `product` argument to
:meth:`~pymor.vectorarrays.interface.VectorArray.gramian` and
:meth:`~pymor.vectorarrays.interface.VectorArray.inner`:

.. jupyter-execute::

    G = trivial_basis[:10].gramian(product=fom.h1_0_semi_product)
    R = trivial_basis[:10].inner(V, product=fom.h1_0_semi_product)
    lambdas = np.linalg.solve(G, R)
    V_h1_proj = trivial_basis[:10].lincomb(lambdas.T)
    
    fom.visualize((V, V_h1_proj, V - V_h1_proj), separate_colorbars=True)

As you might have guessed, we have additionally opted here to only use the
span of the first 10 basis vectors of `trivial_basis`. Like |NumPy arrays|,
|VectorArrays| can be sliced and indexed. The result is *always* a
:attr:`view <pymor.vectorarrays.interface.VectorArray.is_view>` onto the
original data. If the view object is modified, the original array is modified
as well.

Next we will assess the approximation error a bit more thoroughly, by
evaluating it on a validation set of 100 |parameter values| for varying

> basis sized.

Reviewer: ```suggestion
basis sizes.
```

First, we compute the validation snapshots:

.. jupyter-execute::

    validation_set = parameter_space.sample_randomly(100)
    V = fom.solution_space.empty()
    for mu in validation_set:
        V.append(fom.solve(mu))

To optimize the computation of the projection matrix and the right-hand
side for varying basis sizes, we first compute these for the full basis
and then extract appropriate sub-matrices:

.. jupyter-execute::

    def compute_proj_errors(basis, V, product):
        G = basis.gramian(product=product)
        R = basis.inner(V, product=product)
        errors = []
        for N in range(len(basis) + 1):
            if N > 0:
                v = np.linalg.solve(G[:N, :N], R[:N, :])
            else:
                v = np.zeros((0, len(V)))
            V_proj = basis[:N].lincomb(v.T)
            errors.append(np.max((V - V_proj).norm(product=product)))
        return errors

    trivial_errors = compute_proj_errors(trivial_basis, V, fom.h1_0_semi_product)

> Here we have used the fact that we can form multiple linear combinations at once by passing a

Reviewer: ```suggestion
Here we have used the fact that we can form multiple linear combinations at once by passing
```

trivial basis construction algorithm. For compact training sets
contained in a Hilbert space, it can actually be shown that the greedy
algorithm constructs quasi-optimal spaces in the sense that polynomial
or exponential decay of the N-widths :math:`d_N` yields similar rates
for the worst-case best-approximation errors of the constructed :math:`V_N`.

Orthogonalization required
--------------------------

There is one technical problem with both algorithms however: the
condition numbers of the Gramians used to compute the projection
onto :math:`V_N` explode:

.. jupyter-execute::

    G_trivial = trivial_basis.gramian(fom.h1_0_semi_product)
    G_greedy = greedy_basis.gramian(fom.h1_0_semi_product)
    trivial_conds, greedy_conds = [], []
    for N in range(1, len(U)):
        trivial_conds.append(np.linalg.cond(G_trivial[:N, :N]))
        greedy_conds.append(np.linalg.cond(G_greedy[:N, :N]))
    plt.figure()
    plt.semilogy(range(1, len(U)), trivial_conds, label='trivial')
    plt.semilogy(range(1, len(U)), greedy_conds, label='greedy')
    plt.legend()
    plt.show()

This is quite obvious as the snapshot matrix `U` becomes more and
more linear dependent the larger it grows.

> If we would use the bases we just constructed to build a reduced order model

Reviewer 2: ```suggestion
If we would use the bases we just constructed to build a reduced-order model
```

.. math::

    \sum_{k = 1}^K \inf_{v \in V_N} \|u(\mu_k) - v\|^2 =\,
    \min_{\substack{W_N \subseteq V\\ \operatorname{dim}(W_N) \leq N}} \sum_{k = 1}^K \inf_{v \in W_N} \|u(\mu_k) - v\|^2 =
    \sum_{i = N+1}^{r} \sigma^2

Thus, the linear spaces produced by the POD are actually optimal, albeit in a different
error measure: instead of looking at the worst-case best-approximation error over all
|parameter values|, we minimize the :math:`\ell^2`-sum of all best-approximation errors. 
So in the mean squared average, the POD spaces are optimal, but there might be |parameter values|
for which the best-approximation error is quite large.

So far we completely neglected that the snapshot vectors may lie in a Hilbert space
with some given inner product. To account for that, instead of the snapshot matrix
:math:`A`, we consider the linear mapping that sends the :math:`i`-th canonical basis vector
:math:`e_k` of :math:`\mathbb{R}^K` to the vector :math:`u(\mu_k)` in the
:attr:`~pymor.models.interface.Model.solution_space`:

.. math:: \Phi: \mathbb{R}^K \to V, \ e_k \mapsto u(\mu_k).

Also for this finite-rank (hence compact) operator there exists a SVD of the form

> .. math:: \Phi(v) = \sum_{i=1}^r u_i \cdot \sigma_i \cdot (v_i, u) \qquad \forall v \in \mathbb{R}^K,

Reviewer 2: ```suggestion
.. math:: \Phi(v) = \sum_{i=1}^r u_i \cdot \sigma_i \cdot (v_i, v) \qquad \forall v \in \mathbb{R}^K,
```

Now, let us compare how the POD performs against the greedy algorithm in the worst-case
best-approximation error:

.. jupyter-execute::

    pod_errors = compute_proj_errors_orth_basis(pod_basis, V, fom.h1_0_semi_product)
    
    plt.figure()
    plt.semilogy(trivial_errors, label='trivial')
    plt.semilogy(greedy_errors, label='greedy')
    plt.semilogy(pod_errors, label='POD')
    plt.ylim(1e-1, 1e1)
    plt.legend()
    plt.show()

As it turns out, the POD spaces perform even slightly better than the greedy spaces.
Why is that the case? Note that for finite training or validation sets, both considered error measures
are equivalent. In particular:

.. math:: 
    \sup_{k = 1,\ldots,K} \inf_{v \in V_N} \|u(\mu_k) - v\| \leq
    \left[\sum_{k = 1}^K \inf_{v \in V_N} \|u(\mu_k) - v\|^2\right]^{1/2} \leq
    \sqrt{K} \cdot \sup_{k = 1,\ldots,K} \inf_{v \in V_N} \|u(\mu_k) - v\|.

Since POD spaces are optimal in the :math:`\ell^2`-error, they miss the error of the
optimal space in the Kolmogorov sense by at most a factor of :math:`\sqrt{K}`, which in
our case is :math:`5`. On the other hand, the greedy algorithm also only produces

> quasi-optimal spaces. -- For very large training sets with more complex parameter dependence,

Reviewer: ```suggestion
quasi-optimal spaces. -- For very large training sets with more complex parameter dependences,
```

Author: I prefer singular.

Reviewer: I'm fine with either way.

differences between both spaces may be more significant.

Finally, it is often insightful to look at the POD modes themselves:

.. jupyter-execute::

    fom.visualize(pod_basis)

As you can see, the first (more important) basis vectors account for the approximation of
the solutions in the bulk of the subdomains, whereas the higher modes are responsible for
approximating the solutions at the subdomain interfaces.

.. _weakgreedy:

Weak greedy algorithm
---------------------

Both POD and the strong greedy algorithm require the computation of all 
:meth:`solutions <pymor.models.interface.Model.solve>` :math:`u(\mu)`
for all |parameter values| :math:`\mu` in the `training_set`. So it is
clear right from the start that we cannot afford very large training sets.
Otherwise we would not be interested in model order reduction
in the first place. This is a problem when the number of |parameters|
increases and/or the solution depends less uniformly on the |parameters|.

Reduced basis methods have a very elegant solution to this problem, which

> allows training sets that are orders of magnitude larger than the trainings

Reviewer: ```suggestion
allows training sets that are orders of magnitude larger than the training
```

sets affordable for POD: instead of computing the best-approximation error

> we only compute surrogate

Reviewer 3: ```suggestion
we only compute a surrogate
```

.. math:: \inf_{v \in V_N} \|u(\mu) - v\| \approx \mathcal{E}(\mu)

for it. Replacing the best-approximation error by this surrogate in the :ref:`stronggreedy`

> algorithm, we arrive at the :meth:`weak greedy <pymor.algorithms.greedy.weak_greedy>`

Reviewer 2: Right now,
```
:ref:`stronggreedy` algorithm
```
is converted to "Strong greedy algorithm algorithm".

Author: Fixed.

algorithm. If the surrogate :math:`\mathcal{E}(\mu)` is an upper and lower bound
to the best-approximation error up to some fixed factor, it can still be shown that the
produced reduced spaces are quasi-optimal in the same sense as for the strong greedy
algorithm, although the involved constants might be worse, depending on the
efficiency of :math:`\mathcal{E}(\mu)`.

Now here comes the trick: to get a surrogate that can be quickly computed, we can

> use our current reduced order model for it. More precisely, we choose :math:`\mathcal{E}(\mu)`

Reviewer 2: ```suggestion
use our current reduced-order model for it. More precisely, we choose :math:`\mathcal{E}(\mu)`
```

to be of the form

.. math:: \mathcal{E}(\mu):= \operatorname{Err-Est}(\operatorname{ROM-Solve}(\mu), \mu).

So to compute the surrogate for fixed |parameter values| :math:`\mu`, we first

> :meth:`~pymor.models.interface.Model.solve` the reduced order model for the current

Reviewer 2: ```suggestion
:meth:`~pymor.models.interface.Model.solve` the reduced-order model for the current
```

reduced basis for these |parameter values| and then compute an estimate for the
model order reduction error.

We won't go into any further details in this tutorial, but for nice problem classes
(linear coercive problems with an affine dependence of the system matrix on the |Parameters|),
one can derive a posteriori error estimators for which the equivalence with the best-approximation
error can be shown and which can be computed efficiently, independently from the size
of the `fom`. Here we will only give a simple example how to use the 
:meth:`weak greedy <pymor.algorithms.greedy.weak_greedy>` algorithm for our problem at hand.

In order to do so, we need to be able to build a reduced-order

> model with an appropriate error estimator. For the given thermal block problem 

Reviewer 3: ```suggestion
model with an appropriate error estimator. For the given (linear coercive) thermal block problem 
```

a function which estimates the coercivity constant of the system matrix with respect to
the given inner product. In our case, this is just the minimum of the diffusivities over
all subdomains.

Now we can call :meth:`~pymor.algorithms.greedy.rb_greedy`, which constructs for us the
surrogate :math:`\mathcal{E}(\mu)` from the `fom` and the `reductor` we just
constructed. It then passes this surrogate to the :meth:`~pymor.algorithms.greedy.weak_greedy`
method. Furthermore, we need to specify the number of basis vectors we want to compute
(we could also have specified an error tolerance) and the training set.
As the surrogate is cheap to evaluate, we choose here a training set of 1000 different
|parameter values|:

.. jupyter-execute::

    greedy_data = rb_greedy(fom, reductor, parameter_space.sample_randomly(1000),
                            max_extensions=25)

Take a look at the log output to see how the basis is built iteratively using
the surrogate :math:`\mathcal{E}(\mu)`.

The returned `greedy_data` dictionary contains various information about the run
of the algorithm, including the final ROM. Here, however, we are interested in the

> generated reduced basis, which is manged by the `reductor`:

Reviewer: ```suggestion
generated reduced basis, which is managed by the `reductor`:
```

for the worst-case best-approximation errors of the constructed :math:`V_N`.

Orthonormalization required

> --------------------------

Reviewer: ```suggestion
---------------------------
```
